In [1]:
import pandas as pd
import numpy as np

from sqlalchemy import create_engine

In [2]:
engine = create_engine(
    
     "postgresql://postgres:postgres123@localhost:5432/malaria_db"
)

In [3]:
query = "SELECT * FROM malaria_feature_engineered"

df = pd.read_sql(query, engine)

print(df.columns.tolist())

['participant_id', 'dose_num', 'dose_ordered', 'dose_start', 'dose_start_time', 'dose_end', 'dose_end_time', 'dose_complete', 'treatment_duration_days', 'total_doses', 'completed_protocol', 'incomplete_protocol', 'extended_treatment', 'total_exposure', 'avg_dose_ordered', 'avg_duration', 'max_duration', 'missing_end_date', 'missing_duration']


In [4]:
kpi_df = df.copy()

In [5]:
df.shape
df.head()

,participant_id,dose_num,dose_ordered,dose_start,dose_start_time,dose_end,dose_end_time,dose_complete,treatment_duration_days,total_doses,completed_protocol,incomplete_protocol,extended_treatment,total_exposure,avg_dose_ordered,avg_duration,max_duration,missing_end_date,missing_duration
0,1,1,250.0,2019-04-05,TRUE,2019-04-05,TRUE,1,0.0,4,1,0,0,1000.0,250.0,0.0,0.0,0,0
1,1,2,250.0,2019-04-06,TRUE,2019-04-06,TRUE,1,0.0,4,1,0,0,1000.0,250.0,0.0,0.0,0,0
2,1,3,250.0,2019-04-06,TRUE,2019-04-06,TRUE,1,0.0,4,1,0,0,1000.0,250.0,0.0,0.0,0,0
3,1,4,250.0,2019-04-07,TRUE,2019-04-07,TRUE,1,0.0,4,1,0,0,1000.0,250.0,0.0,0.0,0,0
4,2,1,NaN,2019-04-11,TRUE,2019-04-11,FALSE,1,0.0,4,1,0,0,660.0,220.0,0.0,0.0,0,0


# KPI 1 : Total participants


## Objective

Determine the total number of unique participants included in the treatment program.

This KPI establishes the study population size and serves as the denominator for participant-level adherence, exposure, and completion metrics.

## Formula

Total Participants = Count of Unique Participant IDs

In [6]:
total_participants = df['participant_id'].nunique()

total_participants

195

In [7]:
df['participant_id'].nunique()

195

## KPI Result

Total Participants = 195

## Insight

A total of 195 unique participants were included in the treatment dataset.

This population represents the basis for all participant-level healthcare metrics, including treatment adherence, protocol completion, medication exposure, and treatment duration analyses.

All subsequent participant-level KPIs will be calculated using this population as the reference denominator.

# KPI 2 : Treatment Completion Rate

## Objective

Determine the percentage of participants who successfully completed the standard treatment protocol.

This KPI measures treatment adherence by identifying participants who received the expected minimum number of doses required for protocol completion. It serves as a key indicator of program effectiveness and participant compliance.

## Formula
Treatment Completion Rate=
Total Participants
Participants with total_doses≥4
	​

×100

In [8]:
completion_rate= (
df.groupby('participant_id')
      ['completed_protocol']
.first()
.mean()
*100
)

completion_rate

np.float64(75.8974358974359)

## kpi Result

Treatment completion rate = 75.9%

## Insight

A total of 148 out of 195 participants completed the standard treatment protocol, resulting in a treatment completion rate of 75.9%.

This indicates that the majority of participants adhered to the expected treatment regimen and received the minimum four doses required for protocol completion. However, 24.1% of participants did not complete the treatment protocol, suggesting opportunities to improve adherence and ensure participants receive the full intended course of treatment.

Overall, the completion rate demonstrates relatively strong treatment adherence within the study population while highlighting a subgroup of participants who may require additional follow-up or support.

## Conclusion
The treatment program achieved a completion rate of 75.9%, indicating generally good adherence to the prescribed treatment protocol. Monitoring and addressing factors contributing to non-completion could further improve treatment outcomes and program effectiveness.

# KPI 3 : Incomplete Treatment Rate
## Objective

Determine the percentage of participants who failed to complete the standard treatment protocol.

This KPI measures treatment non-adherence by identifying participants who received fewer than four doses. It helps evaluate program retention, adherence challenges, and potential barriers to successful treatment completion.


## Formula
Incomplete Rate = ParticipantsParticipants with total_doses<4×100

In [9]:
incomplete_rate= (
df.groupby('participant_id')
      ['incomplete_protocol']
.first()
.mean()
*100
)

incomplete_rate

np.float64(24.102564102564102)

## kpi Result

incoplete treatment rate = 24.1%

## Insight

A total of 47 out of 195 participants did not complete the standard treatment protocol, resulting in an incomplete treatment rate of 24.1%.

This indicates that nearly one-quarter of participants failed to receive the minimum four doses required for protocol completion. While the majority of participants adhered to the treatment regimen, the observed non-completion rate highlights a potential adherence gap that may affect treatment effectiveness and overall program performance.

Further investigation into the factors contributing to treatment non-completion may help improve adherence and optimize treatment outcomes.

## Conclusion

The incomplete treatment rate of 24.1% suggests that although treatment adherence was generally acceptable, a substantial proportion of participants did not complete the prescribed treatment course. Reducing treatment non-completion should be considered a priority for improving overall program effectiveness.

# KPI 4 : Extended Treatment Rate
## Objective

Extended Treatment Rate measures participants who received more doses than the standard treatment protocol.

## Formula

Extended Treatment Rate=Total ParticipantsParticipants with total_doses>4​×100


In [10]:
extended_rate= (
df.groupby('participant_id')
      ['extended_treatment']
.first()
.mean()
*100
)

extended_rate

np.float64(1.5384615384615385)

# Kpi Result

Extended tratment rate = 1.54%

## Insight

A total of 3 out of 195 participants received more than the standard four-dose treatment protocol, resulting in an extended treatment rate of 1.54%.

This indicates that extended treatment exposure was uncommon within the study population. The low rate suggests strong adherence to the standard treatment protocol and limited occurrence of protocol deviations. Participants receiving additional doses may represent retreatment cases, repeated treatment cycles, or records requiring further clinical review.

## Conclusion

The extended treatment rate of 1.54% demonstrates that treatment beyond the standard protocol was rare. This finding suggests that protocol adherence was generally maintained, with only a small number of participants requiring additional treatment exposure.





# KPI 5 : Average Doses Per Participan
## Objective

Determine the average number of doses received per participant during the treatment program.

This KPI measures overall treatment exposure by calculating the average number of administered doses across all participants. It provides insight into treatment utilization and helps assess adherence to the expected treatment regimen.

## Formula
Average Doses per Participant=
Total Participants
∑Total Doses Received by All Participants
	​


In [11]:
avg_doses= (
df.groupby('participant_id')
      ['total_doses']
.first()
.mean()
)

avg_doses

np.float64(3.7435897435897436)

## Kpi Result

Average doses per participant = 3.74 doses

## Insight

Participants received an average of 3.74 doses during the treatment program. This value is close to the expected four-dose treatment protocol, indicating generally strong adherence across the study population.

The slight difference between the observed average and the expected four doses reflects the presence of participants who did not complete the treatment regimen, as well as a small number of participants who received additional doses beyond the standard protocol.

Overall, the average dose count suggests that treatment administration was largely consistent with the intended treatment schedule.

## Conclusion

The average of 3.74 doses per participant indicates that treatment exposure was generally aligned with the standard four-dose protocol. While adherence was relatively high across the cohort, a small proportion of participants either received fewer than the expected doses or underwent extended treatment, slightly reducing the overall average.

# KPI 6 : Average Participant Exposure
## Objective

Determine the average cumulative medication exposure received by participants during the treatment program.

This KPI measures the overall amount of medication administered to participants and provides insight into treatment intensity, medication utilization, and exposure patterns across the study population.

## Formula

Average Participant Exposure=Total Participants∑total_exposure​

In [12]:
avg_exposure= (
df.groupby('participant_id')
      ['total_exposure']
.first()
.mean()
)

avg_exposure

np.float64(612.2608199666666)

## KPI Result

Average Participant Exposure = 612.26 units



## Insight

Participants received an average cumulative medication exposure of 612.26 units during the treatment program. This indicates the typical amount of medication administered to a participant across all treatment doses.

Exposure levels varied considerably among participants, with total exposure ranging from 94 units to 1444 units. This variation reflects differences in prescribed dose quantities, treatment completion status, and the occurrence of extended treatment among a small number of participants.

Overall, the observed average exposure suggests that participants received substantial medication coverage consistent with the treatment protocol.

## Conclusion

The average participant exposure of 612.26 units indicates that participants generally received substantial cumulative medication throughout the treatment program. While exposure levels varied across individuals, the overall distribution suggests that medication administration was broadly consistent with expected treatment practices. This KPI provides an important measure of treatment intensity and overall medication utilization within the study population.

# KPI 7 : Maximum Participant Exposure
## Objective

Identify the highest cumulative medication exposure recorded among all participants.

This KPI measures the maximum amount of medication received by a single participant during the treatment program and helps identify high-exposure cases that may require further clinical, operational, or protocol compliance review.

## Formula

Maximum Participant Exposure=max(total_exposure)

In [13]:
max_exposure= (
df.groupby('participant_id')
      ['total_exposure']
.first()
.max()
)

max_exposure

np.float64(1444.0)

## KPI Result

Maximum Participant Exposure = 1444 units

## Insight

The highest cumulative medication exposure observed in the study population was 1444 units, recorded for participant 142.

This value is substantially higher than the average participant exposure of 612.26 units, indicating that a small number of participants received considerably greater medication exposure than the typical participant. The elevated exposure may be associated with higher prescribed doses, completion of multiple treatment cycles, or extended treatment beyond the standard protocol.

The presence of high-exposure participants highlights the importance of monitoring cumulative medication exposure to identify unusual treatment patterns and ensure protocol compliance.

## Conclusion

The maximum participant exposure of 1444 units demonstrates that medication utilization varied across the study population. While most participants received exposure levels closer to the cohort average, a small number experienced substantially higher cumulative exposure. Monitoring these high-exposure cases can support treatment oversight, protocol adherence assessments, and exposure-related risk evaluation.

# KPI 8 : Average Ordered Dose


## Objective

Determine the average medication dose ordered for participants during the treatment program.

This KPI measures the typical dose quantity prescribed across participants and provides insight into treatment intensity, prescribing patterns, and medication utilization within the study population.

## Formula

Average Ordered Dose=
Total Participants
∑Average Dose Ordered for All Participants
	​


In [14]:
avg_dose= (
df.groupby('participant_id')
      ['avg_dose_ordered']
.first()
.mean()
)

avg_dose

np.float64(166.00731615864956)

## KPI Result

Average Ordered Dose = 166.01 units

## Insight

Participants received an average ordered dose of 166.01 units throughout the treatment program. This value represents the typical medication quantity prescribed per treatment administration across the study population.

Dose quantities varied among participants, reflecting individualized treatment practices and differences in prescribed medication amounts. Despite this variation, the average dose remained relatively consistent, suggesting that prescribing practices generally followed established treatment guidelines.

The observed dosing pattern indicates that medication administration was tailored to participant needs while maintaining overall consistency across the treatment program.

## Conclusion

The average ordered dose of 166.01 units indicates a generally consistent prescribing pattern across the study population. While individual dose quantities varied, the overall dosing distribution suggests that medication administration was broadly standardized and aligned with expected treatment practices.

# KPI 9 : Average Treatment Duration
## Objective

Determine the average treatment duration experienced by participants during the treatment program.

This KPI measures the typical length of time between treatment initiation and treatment completion and provides insight into treatment timing, operational efficiency, and adherence to expected treatment schedules.

## Formula

Average Treatment Duration=
Participants with Duration Data
∑Average Duration of All Participants
	​


In [15]:
avg_duration= (
df.groupby('participant_id')
      ['avg_duration']
.first()
.mean()
)

avg_duration

np.float64(0.04272151898734177)

## KPI Result

Average Treatment Duration = 0.043 days

## Insight

Participants experienced an average treatment duration of 0.043 days, indicating that treatment administration was typically completed on the same day.

The median treatment duration was zero days, and at least 75% of participants had an average duration of zero days, demonstrating highly standardized treatment delivery. Only a small number of participants experienced extended treatment durations, resulting in a slightly elevated average.

These findings suggest that treatment administration was generally efficient and consistent across the study population.

## Conclusion

The average treatment duration of 0.043 days indicates that treatment was predominantly completed within a single day for most participants. Duration variability was minimal and driven by a small number of prolonged treatment events, suggesting that treatment delivery was generally standardized and operationally efficient.

# KPI 10 : Duration Anomaly Rate
## Objective

Determine the percentage of participants who experienced treatment durations longer than zero days.

This KPI measures the frequency of prolonged treatment events and helps identify unusual treatment patterns, operational delays, or records requiring further clinical review.

## Formula

Duration Anomaly Rate=Participants with Duration DataParticipants with max_duration>0​×100

In [16]:
duration_anomaly_rate= (
    (
df.groupby('participant_id')
          ['max_duration']
.first()
.dropna()
.gt(0)
.mean()
    )*100
)

duration_anomaly_rate

np.float64(4.430379746835443)

## KPI Result

Duration Anomaly Rate = 4.43%

## Insight

Only 7 out of 158 participants with available duration information experienced treatment durations greater than zero days, resulting in a duration anomaly rate of 4.43%.

This indicates that prolonged treatment events were uncommon within the study population. Most participants completed treatment on the same day,

## Conclusion

The duration anomaly rate of 4.43% indicates that prolonged treatment durations were rare within the study population. Most participants completed treatment on the same day, suggesting consistent treatment delivery and adherence to expected treatment timelines. The small number of duration anomalies may warrant further review to determine whether they reflect clinical circumstances, operational delays, or documentation issues.

# KPI 11 : End_Date Documentation Completeness
## Objective

Determine the percentage of treatment records that contain a documented treatment end date.

This KPI measures the completeness of treatment documentation and evaluates the availability of information required for treatment duration analysis and clinical reporting.


## Formula

End-Date Documentation Completeness=
Total Records
Records with End Date
	​

×100

In [17]:
end_date_completeness= (
    (df['missing_end_date']==0)
.mean()
*100
)

end_date_completeness

np.float64(74.10958904109589)

## KPI Result

End-Date Documentation Completeness = 74.1%

## Insight

A total of 541 out of 730 treatment records contained a documented treatment end date, resulting in a documentation completeness rate of 74.1%.

This indicates that most treatment records contained the information necessary for treatment duration calculations and follow-up analyses. However, approximately one-quarter of records lacked end-date documentation, limiting the completeness of duration-based assessments and operational reporting.

## Conclusion

The end-date documentation completeness rate of 74.1% indicates generally acceptable documentation quality. Nevertheless, improving end-date recording practices would enhance the accuracy of treatment duration analyses and support more reliable healthcare reporting.

# KPi 12 : Missing End_Date Rate
## Objective

Determine the percentage of treatment records that lack a documented treatment end date.

This KPI measures documentation gaps and helps identify areas where data completeness can be improved to support accurate treatment monitoring and reporting.

# Formula

Missing End-Date Rate=Total RecordsRecords with Missing End Date​×100

In [18]:
missing_end_date_rate= (
df['missing_end_date']
.mean()
*100
)

missing_end_date_rate

np.float64(25.89041095890411)

## KPI Result

Missing End-Date Rate = 25.9%

## Insight

A total of 189 out of 730 treatment records lacked a documented treatment end date, resulting in a missing end-date rate of 25.9%.

This indicates that approximately one in every four records contained incomplete treatment documentation. Missing end dates reduce the ability to calculate treatment durations and may affect the reliability of duration-related analyses and operational performance metrics.

## Conclusion

The missing end-date rate of 25.9% highlights a notable documentation gap within the dataset. Strengthening end-date recording practices would improve data quality and enhance the accuracy of treatment monitoring and reporting processes.

# KPI 13 : Duration Data Completeness
## Objective

Determine the percentage of treatment records with sufficient information to calculate treatment duration.

This KPI measures the availability of duration-related data and assesses the reliability of treatment timing analyses within the healthcare program.

## Formula

Duration Data Completeness=Total RecordsRecords with Duration Data​×100

In [19]:
duration_completeness= (
    (df['missing_duration']==0)
.mean()
*100
)

duration_completeness

np.float64(74.10958904109589)

## KPI Result

Duration Data Completeness = 74.1%

## Insight

A total of 541 out of 730 treatment records contained sufficient information to calculate treatment duration, resulting in a duration data completeness rate of 74.1%.

The remaining 25.9% of records lacked duration information due to missing treatment end dates. As a result, duration-related analyses could only be performed on records with complete treatment timing information.

## Conclusion

The duration data completeness rate of 74.1% indicates that most records support treatment duration analysis. However, missing duration information in approximately one-quarter of records reduces analytical coverage and highlights the need for improved documentation of treatment end dates.

# KPI Summary Table


In [20]:
kpis=pd.DataFrame({
'KPI':[
'Total Participants',
'Completion Rate',
'Incomplete Rate',
'Extended Treatment Rate',
'Average Doses',
'Average Exposure',
'Maximum Exposure',
'Average Dose Ordered',
'Average Duration',
'Duration Anomaly Rate',
'End Date Completeness',
'Missing End Date Rate'
    ],
'Value':[
total_participants,
completion_rate,
incomplete_rate,
extended_rate,
avg_doses,
avg_exposure,
max_exposure,
avg_dose,
avg_duration,
duration_anomaly_rate,
end_date_completeness,
missing_end_date_rate
    ]
})

kpis

,KPI,Value
0,Total Participants,195.000000
1,Completion Rate,75.897436
2,Incomplete Rate,24.102564
3,Extended Treatment Rate,1.538462
4,Average Doses,3.743590
5,Average Exposure,612.260820
6,Maximum Exposure,1444.000000
7,Average Dose Ordered,166.007316
8,Average Duration,0.042722
9,Duration Anomaly Rate,4.430380


# Overall KPI Summary

The treatment program included 195 participants and achieved a treatment completion rate of 75.9%, indicating generally strong adherence to the standard four-dose protocol. The incomplete treatment rate was 24.1%, highlighting opportunities to improve participant retention and treatment adherence.

Participants received an average of 3.74 doses and an average cumulative medication exposure of 612.26 units. Only 1.54% of participants received treatment beyond the standard protocol, suggesting that protocol deviations were uncommon.

From a data quality perspective, end-date documentation completeness and duration data completeness were both 74.1%, while 25.9% of records lacked treatment end dates. Improving documentation practices would enhance the reliability of duration-related analyses and operational reporting.

In [21]:
kpi_df.to_sql(
    "malaria_kpi",
    engine,
    if_exists="replace",
    index=False
)

730

In [22]:
df.shape
df.columns

Index(['participant_id', 'dose_num', 'dose_ordered', 'dose_start',
       'dose_start_time', 'dose_end', 'dose_end_time', 'dose_complete',
       'treatment_duration_days', 'total_doses', 'completed_protocol',
       'incomplete_protocol', 'extended_treatment', 'total_exposure',
       'avg_dose_ordered', 'avg_duration', 'max_duration', 'missing_end_date',
       'missing_duration'],
      dtype='str')